src/transaction_builder.js

```diff
@@ -372,22 +372,19 @@ TransactionBuilder.prototype.sign = function (index, privKey, redeemScript, hash
   var signatureHash = this.tx.hashForSignature(index, signatureScript, hashType)
 
   if (input.scriptType === 'multisig' && input.redeemScript && input.signatures.length !== input.pubKeys.length) {
-    // store signatures locally
-    var _signatures = input.signatures.slice()
-
-    // loop over pubKeys to set their respective signature or set it to OP_0
-    input.signatures = input.pubKeys.map(function (pubKey) {
-      var signature = null
-      _signatures.forEach(function (_signature, _sigIdx) {
-        // check if the signature is not null / false / OP_0 and verify if it belongs to the pubKey
-        if (!signature && _signature && pubKey.verify(signatureHash, _signature)) {
-          // use .splice to remove the signature from the list, so we won't verify it again
-          signature = _signatures.splice(_sigIdx, 1)[0]
-        }
-      })
+    var signatures = []
+
+    input.pubKeys.forEach(function (pubKey, i) {
+      var allowed = input.signatures.slice(i)
 
-      return signature || ops.OP_0
+      allowed.some(function (signature) {
+        if (!pubKey.verify(signatureHash, signature)) return
+
+        signatures[i] = signature
```

Comment:
Stupid battery died,  charger at work. Will follow up tomorrow.
On 10 Mar 2015 8:46 pm, "Ruben de Vries" <notifications@github.com> wrote:

> In src/transaction_builder.js
> <https://github.com/bitcoinjs/bitcoinjs-lib/pull/378#discussion_r26108461>
> :
>
> >
> > -      return signature || ops.OP_0
> > +      allowed.some(function (signature) {
> > +        if (!pubKey.verify(signatureHash, signature)) return
> > +
> > +        signatures[i] = signature
>
> the tests are failing because with this change if a 2of3 is signed key1
> and key2 (or a 2of2 with just key1) you won't have OP_0s for key3, imo
> you should stick to input.signatures = input.pubKeys.map to make sure
> that input.signatures.length == input.pubKeys.length.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitcoinjs/bitcoinjs-lib/pull/378/files#r26108461>.
>